### PR TITLE
Implement timed_thread_scheduler

### LINKED ISSUE #685: make `run_loop` implementation lock-free

### LINKED ISSUE #847: `schedule_after` should start the work `duration` time after the work is *started*
Currently, it starts the work `duration` time after `schedule_after` is called.

### PR DESCRIPTION
**Pull request by maikel**
Depends on #1275 

Fixes #847

Note, scheduling on this scheduler is wait-free*. the intrusive mpsc queue can be used to address issue #685 

* it is wait-free for producers on the fast-path only

### CODE REVIEW

File: include/exec/__detail/intrusive_heap.hpp [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+      m |= m >> 16;
+      m |= m >> 32;
+      return m + 1;
+    }
+  } // namespace detail
+#  endif
+#endif
+
+  template <auto Key, auto Prev, auto Left, auto Right>
+  class intrusive_heap;
```
Comment (koloshmet) [inline include/exec/__detail/intrusive_heap.hpp:54]: I apologize for the invasion, but I worked on a very similar thing and came to the thought, that erasures from heap are too expensive and therefore cancelation is also. Isn't it better to implement intrusive ordered set based on any balanced binary tree? To my shame I never implemented this idea, but maybe you can find it interesting.

Comment (maikel) [reply include/exec/__detail/intrusive_heap.hpp:54]: I think erasure here is O(log N) (I set the time point value to 0 and call bottom-up-heapify). Thats pretty good.

For a real production version of this I have seen that @lewissbaker or @kirkshoop mentioned some hash based priority queue. But I don't know how that works.

Comment (koloshmet) [reply include/exec/__detail/intrusive_heap.hpp:54]: Yes, you're right, I was too caught up with the thought about multiple erasures that time, sorry

Comment (maikel) [reply include/exec/__detail/intrusive_heap.hpp:54]: Note that this is not a generic ordered set. Deleting an element modifies the key value which is unusual. More over it assumes that a zero initialized key value yields a minimum. More over bottom up heapify will pass the node up even if the parent has a zero timeout, too. That way it is ensured that after calling bottom up the node to be erased is at root and can be popped back.

This is a data structure that works for timeouts but is probably not good for a generic setting

File: include/exec/timed_thread_scheduler.hpp [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+      timed_thread_schedule_operation_base(
+        std::chrono::steady_clock::time_point time_point,
+        void (*set_stopped)(timed_thread_operation_base*) noexcept,
+        void (*set_value)(timed_thread_operation_base*) noexcept) noexcept
+        : timed_thread_operation_base{set_value, command_type::schedule}
+        , time_point_{time_point}
+        , set_stopped_{set_stopped} {
+      }
+
+      std::chrono::steady_clock::time_point time_point_;
```
[the remark this answers is not in the document]
Comment (maikel) [reply include/exec/timed_thread_scheduler.hpp:58]: Is that so.. hm yes I see how thats currently not guaranteed. I will address that

**Comment by maikel** [comment]
I think the current failure in TSAN is due to https://github.com/google/sanitizers/issues/1259 and would be solved if we upgrade gcc to version 12. That works at least locally for me.

I still have to investigate why the tests on windows are failing.

File: include/exec/timed_thread_scheduler.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (maikel) [reply include/exec/timed_thread_scheduler.hpp]: Please have another look :pray: I've added the `when_type` But I only convert the time point to the when type when the operation is started and not at construction. Is this the correct timing to increase the counter?

Should I somehow address the issue that the static counter wraps around or should I assume that `std::size_t` is practically large enough to never wrap around?

[the remark this answers is not in the document]
Comment (maikel) [reply include/exec/timed_thread_scheduler.hpp]: I think I found a better place for the counter. The worker thread of the context can do it without any synchronization. The order of the mpsc queue will define the fifo order. The when type will be created right before inserting it into the internal heap data structure

**Comment by ericniebler** [comment]
c

**Pull request merged**